integration-tests/msa/createMsa.test.ts

```diff
@@ -1,69 +1,244 @@
+                name: 'KeyAlreadyRegistered',
+            })
+        });
+
+        it("should allow new keypair to act for/on MSA", async function () {
+            const additionalKeys = createKeys();
+            const newPayload = {
+                ...payload,
+                newPublicKey: additionalKeys.publicKey,
+            }
+            addKeyData = ExtrinsicHelper.api.registry.createType("PalletMsaAddKeyData", newPayload);
+            ownerSig = signPayloadSr25519(authorizedKeys[0], addKeyData);
+            newSig = signPayloadSr25519(additionalKeys, addKeyData);
+            const op = ExtrinsicHelper.addPublicKeyToMsa(authorizedKeys[0], ownerSig, newSig, newPayload);
+            const events = await op.fundAndSend();
+            assert.notEqual(events["msa.PublicKeyAdded"], undefined, 'should have added public key');
+            authorizedKeys.push(additionalKeys);
+        });
+    });
 
-        const retireMsaEvents = await signAndSend(() => api.tx.msa.retireMsa(), keys);
+    describe("retire/delete keys", function () {
+        let providerKeys: KeyringPair;
+
+        before(async function () {
+            providerKeys = await createAndFundKeypair();
+            const createMsaOp = ExtrinsicHelper.createMsa(providerKeys);
+            await createMsaOp.fundAndSend();
+
+            const createProviderOp = ExtrinsicHelper.createProvider(providerKeys, 'Test Provider');
+            await createProviderOp.fundAndSend();
+        });
+
+        it("should disallow retiring MSA belonging to a provider", async function () {
+            const retireOp = ExtrinsicHelper.retireMsa(providerKeys);
+            await assert.rejects(retireOp.fundAndSend(), {
+                name: 'RpcError',
+                message: /Custom error: 2/,
+            });
+        });
+
+        it("should disallow retiring an MSA with more than one key authorized", async function () {
+            const retireOp = ExtrinsicHelper.retireMsa(keys);
+            await assert.rejects(retireOp.fundAndSend(), {
+                name: 'RpcError',
+                message: /Custom error: 3/,
+            });
+        });
+
+        it("should fail to delete public key for self", async function () {
+            const op = ExtrinsicHelper.deletePublicKey(keys, keys.publicKey);
+            await assert.rejects(op.fundAndSend(), {
+                name: 'RpcError',
+                message: /Custom error: 4/,
+            });
+        });
+
+        it("should fail to delete key if not authorized for key's MSA", async function () {
+            const op = ExtrinsicHelper.deletePublicKey(providerKeys, keys.publicKey);
+            await assert.rejects(op.fundAndSend(), {
+                name: 'RpcError',
+                message: /Custom error: 5/,
+            });
+        });
+
+        it("should test for 'NoKeyExists' error");
+
+        it("should delete all other authorized keys", async function () {
+            for (const key of authorizedKeys) {
```

Comment:
If we remove the `authorizedKeys` array, we don't need to loop here. Not sure it serves a purpose.

Reply:
In earlier tests, we explicitly authorize at least 2 additional keys beyond the "creator" keypair. We need to delete those 2 additional keys before the next test will allow us to retire the MSA. Rather than keeping 2 specific variables for the keypairs added, I think an array is a better/more generic method.

Reply:
If I'm not mistaken, it sounds like we've create the need for this loop by having an array in the first place. If we don't need that array, we won't need this loop either. Is that correct?

Reply:
We somehow need to track all of the keys added to this MSA. Right now there are only 2. But future tests could add more, maybe. An array seemed like the easiest/most generic way to keep track of them. This way, no matter how many keys are added to the MSA, as long as they're added to the array as well, cleanup will take place & subsequent tests will pass.

Reply:
Here's the flow:
1. Create keypair & MSA
2. Add a 2nd keypair to MSA, push to array [array size: 1]
3. Use 2nd keypair to add a 3rd keypair to MSA, push 3rd keypair to aray [array size: 2]
4. Attempt to retire MSA, rejected due to multiple keypairs authorized
5. Loop through array and delete additional keypairs from MSA
6. Now able to successfully retire MSA

Reply:
👍